### ISSUE TITLE
warn_unreachable false positive when reassigning in loop

### ISSUE THREAD

**Issue by Garrett-R** [python/mypy#15917]
**Bug Report**

**To Reproduce**

-Run this code with `warn_unreachable` enabled:

```python
x: int|None = None
x = None
for _ in range(2):
    if x is None:
        print('None')
        x = 2
    else:
        print('Not None')
```
([Mypy playground](https://mypy-play.net/?mypy=latest&python=3.11&flags=warn-unreachable&gist=eebb82e6de21417cbc24c43e75dca970))

**Expected Behavior**

No errors.

**Actual Behavior**

```
main.py:8: error: Statement is unreachable  [unreachable]
```

(this is wrong, the script indeed hits line 8 and prints `Not None`)

**Your Environment**

- Mypy version used: 1.5.1
- Mypy configuration options from `mypy.ini` (and other config files):  `warn_unreachable = true`
- Python version used: 3.11.4

**Comment by erictraut** [comment]
I think this may be a duplicate of #14987, #11612, #10634, #8865, and #8721. All of these seem to be manifestations of the same underlying bug.

**Comment by Garrett-R** [comment]
Ah yes, you're right, my bad!  I had tried searching before creating this, but somehow missed all those.  Closing.
